### ISSUE TITLE
`<multibase-prefix>` in CID definition is unnecessary

### ISSUE THREAD

**Issue by maciejhirsz** [multiformats/cid#6]
The spec explicitly states that CID begins with `<multibase-prefix>`, which is actually not true -- not when the CID is sent over the wire in binary form. Even when CID is sent as multibase encoded string, all the implementator of the spec must know is that it's a multibase encoded byte array following the definition:
```
<cidv1> ::= <cid-version><multicodec-packed-content-type><multihash-content-address>
```

There is no reason to mention `<multibase-prefix>` in the spec, for the same reason there is no need to mention `<multihash-code>` in front of `<multihash-content-address>`. The multibase prefix is a concern for the multibase specification, not the CID specification.

Putting it there suggests that the prefix is either appended to the binary form, which is wrong because it's not, or the multibase encoded form, which is redundant information.

**Comment by diasdavid** [comment]
Thank you for bringing this up @maciejhirsz, it raises an interesting question.

If I understand correctly, your take on the spec is that it is suggesting that if the CID is not multibase encoded with, let's say, base58, there is no reason to add a multibase prefix because we can assume that it is in its binary format.

This is true for any wire protocol we control both the receiving and sending side (i.e bitswap), because we can write our parser to assume that the CID is going on the wired as a byte array. 

However, if we were to receive a CID which we know nothing about (i.e stored on disk), we would need to read its first byte to understand which base it is in, which can go from base2 (binary) to baseEmoji. If we were to omit that first byte, then it would be impossible for a parser to know which way to decode. 

My take on this is that we should just be diligent and `multibase.encode(anyNode, base2)` to send them over the wire. This would force us to release a Bitswap 1.2.0 though.

**Comment by Stebalien** [comment]
> This is true for any wire protocol we control both the receiving and sending side (i.e bitswap), because we can write our parser to assume that the CID is going on the wired as a byte array.

Exactly.

> However, if we were to receive a CID which we know nothing about (i.e stored on disk), we would need to read its first byte to understand which base it is in, which can go from base2 (binary) to baseEmoji. If we were to omit that first byte, then it would be impossible for a parser to know which way to decode.

If we don' t know anything about a CID, we won't even know it's a CID (it will just look like data) so the multibase doesn't help us. If the CID is stored in a known machine readable format, the machine readable format will already tell us what base the CID should be in so we don't need it.

Furthermore, offering a *choice* of base for links stored in encoded IPLD nodes is actually really *bad* as changing the base of a node's links could change the CID of the node itself.

---

Basically, from my point of view, there are conceptual CIDs where no multibase is necessary (it's just a pointer to an object with metadata on how to interpret the object) and then ways to encode CIDs. The correct way to encode a CID for use in text is `<multibase><base-encoded-CID>` but that *doesn't* mean they *have* to be encoded the same way in binary formats where the base is predetermined.

**Comment by Stebalien** [comment]
@lgierth also has an interest in this discussion.

**Comment by vmx** [comment]
I think this got fixed with https://github.com/multiformats/cid/pull/28.

**Comment by vmx** [comment]
Please re-open if you don't agree.
